New authentication

jagerman opened the pull request:
This adds replay-protected, one-shot request authentication, intended to replace the legacy request token interface.

Includes tests.

See the comments at the top of sogs/routes/auth.py for details on how it all works.

(Builds on top of #36 )

jagerman committed:
Add request nonce tracking tables and cleanup code

jagerman committed:
Add X-SOGS-* request authentication

This adds replay-protected, one-shot request authentication, intended to
replace the legacy request token interface.

Includes tests.

See the comments at the top of sogs/routes/auth.py for details on how it
all works.

jagerman committed:
Fix acceptance of unpadded base64

We need to go through our own decode_base64 which handles it properly
rather than base64.b64decode which doesn't (among other problems it
has).

jagerman committed:
Test that inner auth headers on batch subreqs get ignored

jagerman committed:
Room: sort ban list

Makes it deterministic, which makes testing easier but also seems like a
useful thing to do for returned values.

jagerman committed:
Make X-SOGS-* auth apply to legacy endpoints

This allows a new client using X-SOGS-* to still invoke legacy endpoints
without having to construct a legacy Authorization sub-header.

majestrate commented on sogs/routes/auth.py:
```diff
@@ -0,0 +1,236 @@
+    abort(Response(msg, status=code, mimetype='text/plain'))
+
+
+def require_user():
```
decorator?

majestrate commented on sogs/routes/auth.py:
```diff
@@ -0,0 +1,236 @@
+
+    # shared_key is hash of a*B || A || B = b*A || A || B where b/B is the server keypair and A is
+    # the session id pubkey.
+    shared_key = nacl.hash.blake2b(
```
i feel like all of the cryptography details should be in crypto.py

jagerman replied:
Added sogs.hashing for the blake2b calls; the crypto details in here, though, seem non-generic enough and with details like working around the flask full_path bug that it makes more sense to leave them here.

majestrate commented on sogs/routes/auth.py:
```diff
@@ -0,0 +1,236 @@
+
+    hasher = nacl.hashlib.blake2b(
+        request.method.encode() + request.path.encode(),
+        digest_size=64,
```
make this 32 bytes instead so the final digest is a bit more palatable

jagerman replied:
Made it 42 instead (so that it's distinct from common 32-byte values like pubkeys, b64-encodes without padding, and isn't so long).

jagerman committed:
Blake2B: use 42-byte digest, 16-byte personality

42-byte digest because:
- 42 is divisible by 3, and thus padding-free when base64 encoded.
- 42 is obviously not a 32- or 64-byte value such as a pubkey or
  signature.
- 42 is a culturally significant number.

The 16 byte personality is because the personality is *always* a 16 byte
value, and while most interfaces automatically null pad it, some don't,
so a fixed 16-byte value is potentially a bit easier to implement.

jagerman committed:
Add example script generating X-SOGS auth headers

jagerman committed:
Add sogs.hashing with nicer blake2b call

blake2b takes the data either as bytes or as something iterable of bytes
or bytes-like arguments, which makes it a bit nicer to use.

Updates auth.py to use it instead of using libsodium calls directly.

Also add test cases for it.

majestrate commented on sogs/hashing.py:
```diff
@@ -0,0 +1,42 @@
+
+
+def blake2b(
+    data: Union[bytes, Iterable[bytes]],
```
what if this was `Iterable` of things implementing the buffer interface or whatever the base type is called that is accepted by hasher.update?

jagerman replied:
That will work fine as is, so just removed the type hint.

jagerman committed:
Add @user_required decorator for auth-required endpoints

jagerman committed:
Add global user ban request rejection

If you authenticate and are a globally banned user the auth code now
returns a 403 immediately rather than needing individual endpoints to
deal with it.
